src/Metric.jl

```diff
@@ -0,0 +1,245 @@
+$\Gamma_{ijk} = \frac{1}{2} \left[g_{kj,i} + g_{ik,j} - g_{ij,k}\right],$
+
+where $g_{ij,k}=\frac{\partial}{\partial x^k} g_{ij}$ is the coordinate
+derivative of the local representation of the metric tensor.
+"""
+function christoffel_symbols_first(M::MetricManifold, x)
+    ∂g = local_metric_jacobian(M, x)
+    n = size(∂g, 1)
+    Γ = similar(∂g, Size(n, n, n))
+    @einsum Γ[i,j,k] = 1/2 * (∂g[k,j,i] + ∂g[i,k,j] - ∂g[i,j,k])
+    return Γ
+end
+
+@doc doc"""
+    christoffel_symbols_second(M::MetricManifold, x)
+
+Compute the Christoffel symbols of the second kind in local coordinates.
+The Christoffel symbols are (in Einstein summation convention)
+
+$\Gamma^{\ell}_{ij} = \Gamma_{ijk} g^{k\ell},$
+
+where $\Gamma_{ijk}$ are the Christoffel symbols of the first kind, and
+$g^{k\ell}$ is the inverse of the local representation of the metric tensor.
+"""
+function christoffel_symbols_second(M::MetricManifold, x)
+    ginv = inverse_local_metric(M, x)
+    Γ₁ = christoffel_symbols_first(M, x)
+    Γ₂ = similar(Γ₁)
+    @einsum Γ₂[i,j,l] = ginv[k,l] * Γ₁[i,j,k]
+    return Γ₂
+end
+
+"""
+    riemann_tensor(M::MetricManifold, x)
+
+Compute the Riemann tensor, also known as the Riemann curvature tensor,
+at the point `x`.
+"""
+function riemann_tensor(M::MetricManifold, x)
+    n = size(x, 1)
+    Γ = christoffel_symbols_second(M, x)
+    ∂Γ = reshape(ForwardDiff.jacobian(x -> christoffel_symbols_second(M, x), x), n, n, n, n) ./ n
+    R = similar(∂Γ, Size(n, n, n, n))
+    @einsum R[i,j,k,l] = ∂Γ[i,k,l,j] - ∂Γ[i,j,l,k] + Γ[i,k,s] * Γ[s,j,l] - Γ[i,j,s] * Γ[s,k,l]
+    return R
+end
+
+"""
+    ricci_tensor(M::MetricManifold, x)
+
+Compute the Ricci tensor, also known as the Ricci curvature tensor,
+of the manifold `M` at the point `x`.
+"""
+function ricci_tensor(M::MetricManifold, x)
+    R = riemann_tensor(M, x)
+    n = size(R, 1)
+    Ric = similar(R, Size(n, n))
+    @einsum Ric[i,j] = R[i,l,j,l]
+    return Ric
+end
+
+"""
+    ricci_curvature(M::MetricManifold, x)
+
+Compute the Ricci scalar curvature of the manifold `M` at the point `x`.
+"""
+function ricci_curvature(M::MetricManifold, x)
+    ginv = inverse_local_metric(M, x)
+    Ric = ricci_tensor(M, x)
+    S = sum(ginv .* Ric)
+    return S
+end
+
+"""
+    gaussian_curvature(M::MetricManifold, x)
+
+Compute the Gaussian curvature of the manifold `M` at the point `x`.
+"""
+gaussian_curvature(M::MetricManifold, x) = ricci_curvature(M, x) / 2
+
+"""
+    einstein_tensor(M::MetricManifold, x)
+
+Compute the Einstein tensor of the manifold `M` at the point `x`.
+"""
+function einstein_tensor(M::MetricManifold, x)
+    Ric = ricci_tensor(M, x)
+    g = local_metric(M, x)
+    ginv = inverse_local_metric(M, x)
+    S = sum(ginv .* Ric)
+    G = Ric - g .* S / 2
+    return G
+end
+
+function solve_exp_ode(M::MetricManifold, x, v, args...; kwargs...)
+    u0 = x
+    du0 = Base.convert(typeof(x), v)
+    tspan = (0.0, 1.0)
+
+    function exp_problem(du, u, p, t)
+        n = size(u, 1)
```

Comment:
What is `n` good for here? How would one treat matrix manifolds here? Would you `vec` them or compute on matrix level? Maybe it's more the `length(u)` that is relevant?

Reply:
Oh, yeah that `n` is unused right now. Tbh, I've not put much thought into matrix manifolds yet, or even to what a metric matrix should mean in that space. I need to figure out the necessary changes for integrating geodesics on embedded manifolds in coordinates of the embedding first.

We've been using `Sphere` as our test case for vector manifolds. Perhaps we should pick a simple non-Euclidean non-group matrix manifold for testing as well.